DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/svd/src/svd.hpp

```diff
@@ -0,0 +1,264 @@
+
+  // copy input matrix to memory that the device can access
+  q.memcpy(input_matrix_device, a_matrix.data(),
+           matrix_count * kInputMatrixSize * sizeof(float))
+      .wait();
+
+  // read input matrix from DDR
+  auto ddr_write_event = q.submit([&](sycl::handler &h) {
+    h.single_task<IDInputMatrixFromDDRToLocalMem>(
+        [=]() [[intel::kernel_args_restrict]] {
+          MatrixReadFromDDRTo2PipesByBlocks<
+              T, cols, rows, kNumElementsPerDDRBurst, InputMatrixPipe, InputMatrixPipe2>(
+              input_matrix_device, matrix_count, repetitions);
+        });
+  });
+
+  // Compute the covariance matrix
+  q.single_task<IDCovarianceMatrixComputation>(
+      fpga_linalg::StreamingNStdCovarianceMatrix<
+          T, rows, cols, kNumElementsPerDDRBurst, InputMatrixPipe,
+          CovarianceMatrixPipe>());
+
+  // Compute the Eigen values and Eigen vectors
+  q.single_task<IDEigenValuesAndVectorsComputation>(
+      fpga_linalg::StreamingEigen<T, cols, fixed_iteration,
+                                  kNumElementsPerDDRBurst, zero_threshold_1e,
+                                  CovarianceMatrixPipe, EigenValuesPipe,
+                                  EigenVectorsPipe, RankDeficientFlagPipe>());
+
+  // Post process. Using eigen values and eigen vector to produce U, S, V matrix
+  q.single_task<IDUSVFromEigens>(
+      USVFromEigens <T, false, rows, cols, kNumElementsPerDDRBurst,
+                            InputMatrixPipe2, EigenValuesPipe, EigenVectorsPipe,
+                            UTempMatrixPipe, SMatrixPipe, VMatrixPipe>{});
+
+  // Orthogonalizing the U matrix by running it through QRD
+  q.single_task<IDOrthogonalizingU>(
+      fpga_linalg::StreamingQRD<T, false, rows, rows, raw_latency,
+                                kNumElementsPerDDRBurst, UTempMatrixPipe,
+                                UMatrixPipe, RMatrixPipe>());
+
+  // Write the rank deficient flag from local memory to FPGA DDR
+  auto rank_deficient_flag_event =
+      q.single_task<IDRankDeficientFlagFromLocalMemToDDR>(
+          [=]() [[intel::kernel_args_restrict]] {
+            VectorReadPipeToDDR<ac_int<1, false>, 1, RankDeficientFlagPipe>(
+                rank_deficient_flag_device, matrix_count, repetitions);
+          });
+
+  // terminating R matrix pipe from the orthogonalizing kernel
+  // This kernel only read from the pipe 
+  // (and throw away the data since its not useful here)
+  q.single_task<IDRMatrixPipeTerminate>(
+      [=]() [[intel::kernel_args_restrict]] {
+        T r_sum = 0.0;
+        for (int mat_idx = 0; mat_idx < matrix_count; mat_idx ++) {
+          for (int rep = 0; rep < repetitions; rep++) {
+            for (int r_idx = 0; r_idx < kRMatrixSize; r_idx++) {
+              r_sum += RMatrixPipe::read();
+            }  // end of r_idx
+          }    // end of rep
+        }
+      });
+
+  // collecting U matrix from pipe into DDR
+  auto u_matrix_event = q.single_task<IDUMatrixFromLocalMemToDDR>(
```

Comment:
```suggestion
  sycl::event u_matrix_event = q.single_task<IDUMatrixFromLocalMemToDDR>(
```

no need for `auto`, please do this for other events